proc_macro: implement `TokenTree`, `TokenKind`, hygienic `quote!`, and other API

jseyfried opened the pull request:
All new API is gated behind `#![feature(proc_macro)]` and may be used with `#[proc_macro]`, `#[proc_macro_attribute]`, and `#[proc_macro_derive]` procedural macros.

More specifically, this PR adds the following in `proc_macro`:
```rust
// `TokenStream` constructors:
impl TokenStream { fn empty() -> TokenStream { ... } }
impl From<TokenTree> for TokenStream { ... }
impl From<TokenKind> for TokenStream { ... }
impl<T: Into<TokenStream>> FromIterator<T> for TokenStream { ... } 
macro quote($($t:tt)*) { ... } // A hygienic `TokenStream` quoter

// `TokenStream` destructuring: 
impl TokenStream { fn is_empty(&self) -> bool { ... } }
impl IntoIterator for TokenStream { type Item = TokenTree; ... }

struct TokenTree { span: Span, kind: TokenKind }
impl From<TokenKind> for TokenTree { ... }
impl Display for TokenTree { ... }

struct Span { ... } // a region of source code along with expansion/hygiene information
impl Default for Span { ... } // a span from the current procedural macro definition
impl Span { fn call_site() -> Span { ... } } // the call site of the current expansion
fn quote_span(span: Span) -> TokenStream;

enum TokenKind {
    Group(Delimiter, TokenStream), // A delimited sequence, e.g. `( ... )`
    Term(Term), // a unicode identifier, lifetime ('a), or underscore
    Op(char, Spacing), // a punctuation character (`+`, `,`, `$`, etc.).
    Literal(Literal), // a literal character (`'a'`), string (`"hello"`), or number (`2.3`)
}

enum Delimiter {
    Parenthesis, // `( ... )`
    Brace, // `[ ... ]`
    Bracket, // `{ ... }`
    None, // an implicit delimiter, e.g. `$var`, where $var is  `...`.
}

struct Term { ... } // An interned string
impl Term {
    fn intern(string: &str) -> Symbol { ... }
    fn as_str(&self) -> &str { ... }
}

enum Spacing {
    Alone, // not immediately followed by another `Op`, e.g. `+` in `+ =`.
    Joint, // immediately followed by another `Op`, e.g. `+` in `+=`
}

struct Literal { ... }
impl Display for Literal { ... }
impl Literal {
    fn integer(n: i128) -> Literal { .. } // unsuffixed integer literal
    fn float(n: f64) -> Literal { .. } // unsuffixed floating point literal
    fn u8(n: u8) -> Literal { ... } // similarly: i8, u16, i16, u32, i32, u64, i64, f32, f64
    fn string(string: &str) -> Literal { ... }
    fn character(ch: char) -> Literal { ... }
    fn byte_string(bytes: &[u8]) -> Literal { ... }
}
```
For details on `quote!` hygiene, see [this example](https://github.com/rust-lang/rust/pull/40939/commits/20a90485c040df87a667e9b6ee38e4d8a7d7fc5d) and [declarative macros 2.0](https://github.com/rust-lang/rust/pull/40847). 

r? @nrc

jseyfried commented:
cc @rust-lang/compiler @abonander @dtolnay @Ericson2314 @SergioBenitez @SimonSapin @sgrif @tikue @wycats

eddyb commented:
Can't we call `TokenTree` just `Token`? I'm a bit disappointed in how I fooled myself we could get rid of the "tree" model but flattening seems to not be in the cards.
*However*, I still think we can keep simpler names in the API.

SimonSapin commented:
Nice! I like the overall approach, some comments on the details:

> ```rust
> impl IntoIterator for TokenStream { type Item = TokenTree; ... }
> ```

Would impls for `&TokenStream` and `&mut TokenStream` also be useful?

> ```rust
> struct Symbol { ... } // An interned string
> ```

To me the word "symbol" evokes punctuation, excluding letters. Could this be called something else? `Identifier`, `Ident`, `Word`, … ? (Are the first two not appropriate because "identifier" excludes keywords?)

> ```rust
> Joint, // immediately followed by another `Op`, e.g. `+` in `+=`
> ```

So the `=` in `+=` would be a separate token? `&&`, `..`, or `::` would similarly be two tokens each? (`>>=` or `...` three tokens?) Please expand the doc-comments in the PR to explain this.

> ```rust
> impl Display for Literal { ... }
> impl Literal {
> ```

I’d also add:

* `fn byte_string(bytes: &[u8]) -> Literal`, much more convenient that building a `Sequence` of `u8` literals for a slice literal. (And more efficient?) 
* Maybe `fn byte(byte: u8) -> Literal` similar to `Literal::u8` but with byte literal syntax, though it only makes a difference when looking at generated code for debugging.
* `fn as_str(&self) -> Option<&str>` and other `fn as_*(&self) -> Option<*>` methods to extract the value. (Though number literals without a type suffix are tricky.) The `Display` impl technically provides the same information, but writing a Rust parser (even just for literals) is a complex undertaking.

---
In `src/libproc_macro/lib.rs`

```rust
#[unstable(feature = "proc_macro", issue = "38356")]
#[macro_export]
macro_rules! quote { () => {} }
```

In `src/libproc_macro/quote.rs`
```rust
macro_rules! quote {
    () => { TokenStream::empty() };
    ($($t:tt)*) => { [ $( quote_tree!($t), )* ].iter().cloned().collect::<TokenStream>() };
}
```

Why are there two `quote` macros, with one empty, and only the empty one exported?

SimonSapin commented:
I think this PR addresses all of my previous comments in https://github.com/rust-lang/rust/issues/38356#issuecomment-276676526 👍

bors commented:
:umbrella: The latest upstream changes (presumably #40950) made this pull request unmergeable. Please resolve the merge conflicts.

jseyfried commented:
@eddyb 
> I'm a bit disappointed in how I fooled myself we could get rid of the "tree" model but flattening seems to not be in the cards.

Do you mean w.r.t. the internals? I believe we can still use the internal flattened representation you proposed with this API (imo, would be nice but not high priority).

@SimonSapin 
> Would impls for &TokenStream and &mut TokenStream also be useful?

`impl IntoIterator for &TokenStream { type Item = TokenTree; ... }` and/or
`impl TokenStream { fn iter(&self) -> TokenIter { ... } }` could be useful -- both would be equivalent to `tokens.clone().into_iter()`.

`impl IntoIterator for &TokenStream { type Item = &TokenTree; ... }` isn't possible (yet, perhaps), and would make an internal "flattened" internal representation more complex / less efficient.

`TokenStream` uses shared, ref-counted memory, so I don't think `impl IntoIterator for &mut TokenStream` makes sense.

> To me the word "symbol" evokes punctuation, excluding letters. Could this be called something else?

Yeah. Currently, `Symbol` is just an interned string, so `Word`/`Ident` don't seem appropriate. If we restrict the type to valid `TokenKind::Word`s, I'd prefer `Word` for symmetry with `TokenKind`. Otherwise, perhaps `Str`/`InternedStr`? cc @nrc 

> So the `=` in `+=` would be a separate token? `&&`, `..`, or `::` would similarly be two tokens each?

Yeah, it's `TokenKind::Op(char, OpKind)`, so a just single character per `Op` -- I'll mention that in the doc comment.

> I’d also add [more methods for `Literal`]

Yeah, I agree that we should have `Literal::{byte, byte_string, as_*}`. This API isn't intended to be exhaustive, e.g. we also want API for emitting "first class" diagnostics (@Sergio is working on this).

> Why are there two `quote` macros, with one empty, and only the empty one exported?

The non-empty one is incomplete (by design) and only used to bootstrap the internal implementation of `proc_macro::quote`. The exported one is a placeholder for the internal implementation (à la lang items). After we support stability hygiene and stability-checking macro resolutions, we should be able to replace the internal implementation with a `#[proc_macro]`.

bors commented:
:umbrella: The latest upstream changes (presumably #40811) made this pull request unmergeable. Please resolve the merge conflicts.

nrc commented on src/libsyntax/tokenstream.rs:
```diff
@@ -196,6 +201,10 @@ impl TokenStream {
         }
     }
 
+    pub fn builder() -> TokenStreamBuilder {
```
I think this would be better as a free function rather than a method on TokenStream

nrc replied:
Or it could be a `new` method on TokenStreamBuilder

jseyfried replied:
I'll change to `TokenStreamBuilder::new()` (motivation for `TokenStream::builder()` was to avoid making people `use` anything else into their scopes).

nrc commented on src/libsyntax/tokenstream.rs:
```diff
@@ -225,13 +234,107 @@ impl TokenStream {
         true
     }
+
+    pub fn as_tree(self) -> (TokenTree, bool /* joint? */) {
```
Could these public methods get docs please? (In particular here, some more explanation of what joint means would be good).

lfairy replied:
Also, since this takes `self` by value, should this be named `into_tree` instead? Or are the conventions different for reference counted things

jseyfried replied:
@nrc will do.
@lfairy Perhaps, I don't think it matters much.

nrc commented on src/test/run-pass-fulldeps/auxiliary/proc_macro_def.rs:
```diff
@@ -8,47 +8,37 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-#![feature(plugin, plugin_registrar, rustc_private)]
```
It looks like anyone with an existing procedural macro is going to have to make some significant changes. Could you write up either here or on internals or somewhere, a guide to the changes needed to make a proc macro work after this PR please?

jseyfried replied:
Existing procedural macros will lose the `syntax::tokenstream::TokenStream` quoter from `proc_macro_internals`, AFAIK other than that they won't have to make any changes. I could include a `syntax::tokenstream::TokenStream` quoter for back-compat.

How many people are actually using "existing procedural macros" though? This PR doesn't effect the widely used legacy plugin system; I didn't think many people were using `SyntaxExtension::ProcMacro` from #36154.

nrc commented on src/libsyntax/parse/token.rs:
```diff
@@ -455,3 +461,38 @@ pub fn is_op(tok: &Token) -> bool {
 }
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct LazyTokenStream(RefCell<Option<TokenStream>>);
```
Is this meant to be used by macro authors? If so could you document with why and when.

jseyfried replied:
No, nothing in `libsyntax` can be used by proc macro authors.

lfairy commented on src/libsyntax/parse/token.rs:
```diff
@@ -395,3 +452,47 @@ impl fmt::Debug for Nonterminal {
+}
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct LazyTokenStream(RefCell<Option<TokenStream>>);
```
As an aside: if the dynamic borrow checks end up too inefficient, then this could be re-implemented on top of [`MoveCell`](https://github.com/SimonSapin/rust-movecell).

SimonSapin replied:
`std::cell::Cell` is now like `MoveCell` :) https://github.com/rust-lang/rfcs/pull/1651

lfairy commented on src/libsyntax/parse/token.rs:
```diff
@@ -395,3 +452,47 @@ impl fmt::Debug for Nonterminal {
+        let mut opt_stream = self.0.borrow_mut();
+        if opt_stream.is_none() {
+            *opt_stream = Some(f());
+        };
```
Nit: extra semicolon

jseyfried commented on src/libproc_macro/lib.rs:
```diff
@@ -69,6 +76,443 @@ pub struct LexError {
+
+/// An iterator over `TokenTree`s.
+#[unstable(feature = "proc_macro", issue = "38356")]
+pub struct TokenIter {
```
@nrc Yeah, `TokenTreeIter` makes more sense.

On a similar note, @eddyb proposed using `TokenNode` instead of `TokenKind` for better correspondence to `TokenTree` and to leave room for `Token`/`TokenKind` -- what do you think?

jseyfried commented on src/libproc_macro/lib.rs:
```diff
@@ -69,6 +76,443 @@ pub struct LexError {
+}
+
+impl TokenTree {
+    fn from_raw(stream: tokenstream::TokenStream, next: &mut Option<tokenstream::TokenStream>)
```
I'll change to `from_internal`.

jseyfried commented on src/libproc_macro/lib.rs:
```diff
@@ -69,6 +76,443 @@ pub struct LexError {
+        TokenTree { span: Span(span), kind: kind }
+    }
+
+    fn to_raw(self) -> tokenstream::TokenStream {
```
It's all the `pub` API not in `__internal` (which is unusable without `#![feature(proc_macro_internals)]`).

jseyfried commented on src/libproc_macro/lib.rs:
```diff
@@ -80,7 +524,11 @@ pub struct LexError {
 /// all of the contents.
 #[unstable(feature = "proc_macro_internals", issue = "27812")]
 #[doc(hidden)]
+#[path = ""]
```
`mod quote;` in `__internal`; this allows the file to be `libproc_macro/quote.rs`. Thinking about this some more, I'll move `mod quote` into the crate root and remove the `#[path = ""]`.

jseyfried commented on src/libproc_macro/quote.rs:
```diff
@@ -0,0 +1,249 @@
+
+//! # Quasiquoter
+//! This file contains the implementation internals of the quasiquoter provided by `qquote!`.
+
```
Will do.

carols10cents commented:
Friendly ping that there are merge conflicts and some comments for you @jseyfried! ❤️

abonander commented:
Should we use a different feature gate for the unstable APIs? I'm currently under the impression that language feature gates and library feature gates should be distinct, and `proc_macro` currently falls more under the language side.

nikomatsakis commented:
Hmm, is there a strict separation between language/lib feature-gates? I wouldn't think so.

abonander commented:
It's kind of implied by the structure of the Unstable Book but that might be more of an issue for #41476 where this came up originally.

arielb1 commented:
status: still waiting for https://github.com/rust-lang/rust/pull/40847.

arielb1 commented:
status: still waiting for #40847.

carols10cents commented:
status: still waiting for #40847.

Mark-Simulacrum commented:
Still waiting.

alexcrichton commented:
Since enduring [awful error messages](https://github.com/alexcrichton/futures-await/#compiler-errors) with my toy async/await idea I now have quite a bit of interest in this :)

Along those lines I was hoping I could start doing the legwork to prepare libraries like `syn` to take advantage of this API, and I figured I could write down some things I'm learning as I go along. So far everything's related to the API of `proc_macro` as a library, and I've got:

* Perhaps the various exported structs should all implement `Debug`? (just a nice convenience)
* For floating point literals we may wish to panic if you pass in NaN, infinity, or negative infinity. I don't think we have literals for those, right?
* The `quote` crate (on crates.io) currently has support for byte string literals and hexadecimal integer literals. I wonder if perhaps this should also have support for those literal types? The byte string one is relatively easy but the hexadecimal integer literals is pretty interesting. It also interacts with how `quote! { #integer_variable }` produces something like `1i32` by default (as opposed to `1`) I think for disambiguating. Should the API here give a level of control to what the stringified integer literal looks like?

Nothing major at all for landing this PR, just wanted to jot down my thoughts!

alexcrichton commented:
It also looks like doc comments are mapped to a `Literal`, but I don't think there's a way to create a `Literal` from a doc comment str?

dtolnay commented:
### Unsuffixed numeric literals

Like Alex mentioned, I think these are important. For example a proc macro containing this:

```rust
quote! {
    f(0);
}
```

Using the current API there would be no way for `quote!` to correctly map this to tokens because we can't know what integer type `f` expects.

Having additional constructors `Literal::integer` and `Literal::float` would solve this.

### Byte and byte-string literals

How else could it handle `quote! { b'\xFF' b"\xFF" }`. There should be Literal constructors for these.

### Boolean words

Are `true` and `false` represented as TokenKind::Word or TokenKind::Literal? When handing tokens back to the compiler, will it accept either one? Also there should be `Literal::boolean` to construct a boolean literal.

### Naming of `Op`

What does `Op` stand for? If "operator", I would not consider things like `,` and `$` to be operators. Is there a better name?
 
### Canonical `quote!` macro

What is the advantage of providing a canonical `quote!` macro in `proc_macro`? No, I am not just asking this because I have a quote crate. It seems out of place in that every other piece of API added here is crucial to passing code into and out of a proc macro. But with the right rest of the API, `quote!` should be possible to implement on top of it. Is there a part I am missing that gives `quote!` superpowers over what would be possible in a crate? Why not open this up to innovation in the crate ecosystem?

### Deref for Symbol

Let's not. There are so many methods on `str` and you would never want to call any of them on a Symbol. Like 6 trim methods and 6 split methods? It pollutes the rustdoc of Symbol and makes it impossible to find the methods you will actually want to use. I would prefer an explicit method to extract a \&str.

jseyfried commented:
@alexcrichton 
> Perhaps the various exported structs should all implement Debug? (just a nice convenience)

Agreed, will do.

> For floating point literals we may wish to panic if you pass in NaN, infinity, or negative infinity. I don't think we have literals for those, right?

Good point. I believe everything works as is, the only issue is asymmetry -- `Token` is more general than a token in the source since the latter can't represent `NaN` as a single `Literal` token.

Perhaps (some of) the `Literal::*` constructors should instead be free functions and that return `TokenStream`?
E.g. `float_literal: f32 -> TokenStream` would map `NaN` to `f32::NAN` (which will always resolve as expected due to hygiene).

> I wonder if perhaps this should also have support for those literal types?

I think so. This API isn't intended to be complete -- there are a variety of orthogonal improvements that we can work on in parallel once this lands.

> It also looks like doc comments are mapped to a `Literal`, but I don't think there's a way to create a `Literal` from a doc comment str?

Yeah, I'm not sure if this is the best way to treat doc comments -- we might instead add a fifth variant `TokenKind::DocComment` (unaesthetic imo since the other four are more fundamental/general) or do something else.

> Should the API here give a level of control to what the stringified integer literal looks like?

Yes, I think we should expose this. I'll add @dtolnay's `Literal::integer` and `Literal::float`.

@dtolnay 

> Byte-string literals

Yeah, this should be in the API.

> Are true and false represented as `TokenKind::Word` or `TokenKind::Literal`?

Right now, `TokenKind::Word` like other keywords. I think this is simpler, since it avoids the `TokenKind::Word("true")` vs `Token::boolean(true)` issue you mentioned and means unicode identifiers are valid `Word`s without exception (I think). That being said, I'm open to a different model.

> Is there a better name [than `Op` / operator]?

Perhaps -- maybe something like `Char`?
IIRC @eddyb proposed `Op` for brevity.

> ... with the right rest of the API, `quote!` should be possible to implement on top of it.

Agreed. With `Literal` API you mentioned earlier and a function `quote_span: Span -> TokenTree`, I believe you could implement `quote!` from the other API. We don't have `quote_span` yet, but it is easy to add.

> Is there a part I am missing that gives `quote!` superpowers over what would be possible in a crate?

The missing piece is quoting spans. If you e.g. just use the dummy span, then hygiene and expansion backtraces won't work (both are determined by `Span`s' expansion information), and it will limit how powerful the compiler error messages can be. I'm not sure exactly what primitives would be best to expose here, but `quote_span: Span -> TokenTree` seems like a good start -- it's straightforward and sufficient to implement `quote!`.

> Why not open this up to innovation in the crate ecosystem?

It is open to the crate ecosystem. `proc_macro::quote!` is just a declarative macro item in `proc_macro`; people are free to `use third_party::quote;`. I'm envisioning that third parties won't "publicly" depend on `proc_macro` at all (i.e. end users won't directly import anything from `proc_macro`).

A third party library could just re-export `proc_macro::quote!` or provide their own, more powerful version.

> What is the advantage of providing a canonical `quote!` macro in proc_macro?

A benefit right now is we can experiment with a basic `quote!` without waiting for third party libraries.

Also, it should be useful for implementing a more powerful `quote!` -- whenever I implement `quote!` I find I need to "bootstrap" it with a limited "stage 0" version for use in the implementation, which is generally tedious and uninteresting code. Overall, I think a basic `quote!` is worth it's weight to provide in `proc_macro`, even if abstractions over `proc_macro` provide their own quote.

> Let's not [Deref for Symbol]

But `Symbol` is just an interned `str` right now -- it could be `Interned<str>` where `Interned<T>` is a general smart pointer. If perf weren't an issue, it'd just be `String` (like how `Op` is `char`).

dtolnay commented:
- For inf and nan - I would prefer Alex's suggestion of panicking when passed to the Literal constructor. I have never had a use case where I would have wanted them to turn into `f32::NAN`.

- I agree, we don't need hex (or in general, human-friendly) literals in this initial implementation.

- In hindsight how do we feel about the macro_rules handling of doc comments?

    ```rust
    macro_rules! doc {
        (# [ doc = $s:tt ]) => { $s }
    }

    fn main() {
        // prints " informative "
        println!("{:?}", doc!(/** informative */));
    }
    ```

    I thought we were good with that. Can we do doc comments that way here? `#` + bracketed content.

    One limitation is there wouldn't be a way to track what is currently `is_sugared_doc`. For example in `stringify!(/** informative */)` which prints as a doc comment, not `#[doc]`. Would a proc macro ever require knowing whether it is a sugared doc? I don't think so, in fact it may be better if they can't.

    If necessary, one possible hackaround would be treat it as Joint(#) + Joint([) + Word(doc) where the `[` is joint instead of alone when sugared.

    I agree TokenKind::DocComment is not great but I could accept that too.

- `Char` instead of `Op` isn't great either. Char means character literal `'c'`.

    Related: I'm also not happy with `Symbol`. Something like `$` is colloquially way more strongly associated with the word "symbol" than "operator".

    I will try to come up with some good alternatives.

- TokenKind::Word for `true` and `false` is great. I agree with the justification you gave.

- If that's all that is missing to empower an external `quote!` implementation, I would include it here. No need to put off something we know we need.

- Regarding Deref for Symbol: I don't find that compelling. I understand it is just a smart pointer to str but that doesn't seem like the intended way to interact with it or conceptualize it. It should be comparatively rare for a proc macro to need the str of a Symbol or to call str methods on it, so there should be an explicit indication that the user meant to do that.

SimonSapin commented:
> unicode identifiers are valid Words without exception (I think)

Non-ASCII identifiers go through some amount of normalization, so `Symbol::from(s).deref() == s` doesn’t hold for every `s: &str`. If that can’t be the case since `Symbol` is also used in other cases, then a `struct Indent(Symbol);` wrapper type doing normalization is needed.

dtolnay commented:
Here are three naming ideas.

```rust
enum TokenKind {
    Group(Delimiter, TokenStream),
    Term(Term),
    Punctuation(char, PunctuationKind),
    Literal(Literal),
}
```

**`Group`**—I like this strictly better than Sequence. People are used to thinking of parentheses in expressions as grouping things. Sequence sounds flat.

**`Term`**—I like this one except for unfortunately it is a prefix of words like terminate, terminal, terminus which may be confusing. I like that the meaning perfectly covers both keywords and names while excluding punctuation symbols.

**`Punctuation`**—This is the right word but lots of characters. Punct could work but is inconsistent because nothing else is abbreviated (after removing Op of course).

dtolnay commented:
Discussed above:

- [ ] Debug for everything
- [ ] Literal constructor panics on bad floats
- [ ] Literal::integer
- [ ] Literal::float
- [ ] Literal::byte_string
- [ ] Naming in TokenKind (my previous comment or similar)
- [ ] Remove Deref for Symbol
- [ ] Treat doc comments as #[doc = ...] instead of Literal
- [ ] quote_span

eddyb commented:
@jseyfried As per previous discussions, we should be wary of float literals. I would prefer it if floats were converted into token streams of multiple tokens even for simple cases such as `1.0`.
There are also literal suffixes that can be split and we could be using spans instead of explicit `Spacing`.
I don't feel too strongly about any of these but ideally we'd move in that direction.

alexcrichton commented:
All the suggestions sound great to me here, thanks for jumping in @dtolnay! 

Also to be clear I'm fine landing this PR ahead of resolving 100% of these questions, it's all unstable anyway! So long as we keep track of @dtolnay's [list](https://github.com/rust-lang/rust/pull/40939#issuecomment-303302410) it seems like it'd be ok to land at least a preliminary version while the fine details are all decided before stabilization.

mystor commented:
A few notes from my small amount of interaction with these APIs:

1. It would be nice to have `Eq`/`PartialEq` implementations on `Delimiter`, and `OpKind`. In addition, having `PartialEq` implementations on `Symbol` for `&str` etc. would be nice for convenience of writing code interacting with this token type.

2. I like @dtolnay's names better than the current ones for the variants in TokenKind - but I'm not super picky.

3. I don't like that Doc Comments are rolled into `Literal`, because it means I can't do things like lazily consume `Literal` tokens while parsing an AST, and then only translate them into a value literal when consumed, as I don't know if the object is legal in an expression position. A mild nit (I can do a to_string().starts_with() check to quickly check if its a Doc Comment). My favorite option would be to do what `dtolnay` was suggesting with translating it to `#[doc="..."]` before generating tokens, but IIRC that isn't backwards compatible. I would prefer a `TokenKind::DocComment` to them being part of `Literal`.

4. It would be awesome if `Literal` acquired some methods for parsing the literal values into actual values, perhaps `parse_u32(&self) -> Option<u32>` to get a `u32`, etc. This should probably be done in a follow-up though. I think that one of the largest unnecessary sources of complexity in libraries like `syn` is the code to interpret a string literal or integer into a useful type.

5. If we change `syn` to use `TokenStream` as its input type instead of `&str`, it will be really awesome for procedural macro writers, but it also will prevent people from running it outside of the compiler. I would _really_ like it if we could make proc-macro run correctly outside of rustc, when a parse session is not available, such as in build scripts. If we can't do that I'll put in the legwork to allow using an alternate tokenstream implementation (like my fauxenstream or @alexcrichton's proc-macro2), but I'd rather not.

alexcrichton commented:
On tracking-related question issue I'd have: where do we want to track all these questions about the `proc_macro` API itself? @jseyfried do you have a preference for whether it's done as part of https://github.com/rust-lang/rust/issues/38356 or a separate tracking issue?

I might personally lean towards a separate tracking issue as I think this crate should continue to grow tons of sweet apis to use for procedural macros, but I think the first pass for stabilization is going to want to happen relatively soon and continue to be quite conservative (e.g. what's exposed in this PR). A separate tracking issue would help clearly draw the line of "more APIs are coming, this is just the first wave of stabilization" and the latter waves of APIs could continue to go in https://github.com/rust-lang/rust/issues/38356. This'd also allow us to categorize APIs as they go into `proc_macro` itself as "likely to be stabilized soon" and "nice-to-have eventually" perhaps.

jseyfried commented:
@dtolnay 
> Can we do doc comments that way [status quo from `macro_rules!`] here?

Yeah, I think that makes the most sense for now, at least until an alternative is proposed (seems like an orthogonal issue that should be its own RFC, actually).

> If that's all that is missing to empower an external `quote!` implementation, I would include it here. 

Will do.

@SimonSapin 
> If that can’t be the case since `Symbol` is also used in other cases, then a `struct Indent(Symbol);` wrapper type doing normalization is needed

Ok, I agree that `Symbol`/`Term` should be normalized. I'm still in favor of `Deref<Target=str>`, but I'll remove it in this initial implementation since it's controversial.

@dtolnay I like those names (`Group`, `Term`, `Spacing`, and `Punctuation`/`Punct`). I think the abbreviating "punctuation" is worth it for more ergonomic `match`s.

@eddyb 
> As per previous discussions, we should be wary of float literals. I would prefer it if floats were converted into token streams of multiple tokens even for simple cases such as `1.0`.

I think we should keep the status quo for this PR, but we could "future proof" the API by providing `quote_f32(value: f32) -> TokenStream` instead of `impl Literal { fn f32(value: f32) -> Self { ... } }`. @dtolnay @nrc thoughts?

@mystor 
> It would be nice to have `Eq`/`PartialEq` implementations on `Delimiter` and `OpKind`.

Will do.

> I would really like it if we could make `proc-macro` run correctly outside of `rustc`, when a parse session is not available, such as in build scripts.

I'm not sure exactly how spans' expansion backtraces and hygiene info would work outside `rustc` -- currently, serializing/deserializing loses this information (see limitation section of #40847). I think we should explore this, but it's outside the scope of this PR.

> @jseyfried do you have a preference for whether [proc macro API discussion is] done as part of #38356 or a separate tracking issue?

I don't have a strong preference -- keeping #38356 focused a "first pass stabilization" and differing longer term goals to other RFCs as you described makes sense.

clarcharr commented:
Is there a particular reason why literals are encapsulated within a struct instead of using a non-exhaustive enum like `io::Error` is? I understand the need to add additional literal types in the future, but the current way of doing it seems incredibly hackish.

alexcrichton commented:
I was testing this out locally and I think I may have found some odd behavior. I had a function like:

```rust
#[async]
fn main() {
    let a: i32 = 0u32;
}
```

If I just return the `function: TokenStream` I'm given (don't transform it at all) then I get:


```
error[E0308]: mismatched types
  --> src/main.rs:17:17
   |
17 |     let b:i32 = 0u32;
   |                 ^^^^ expected i32, found u32

error: aborting due to previous error(s)
```

(as expected)

If I, however, change the function to return `function.into_iter().collect()` (basically just iterate over all tokens and re-collect them back to a `TokenStream`) then I get:

```
error[E0308]: mismatched types
 --> <macro expansion>:1:1
  |
1 | fn main() { let b: i32 = 0u32; }
  | ^^ expected i32, found u32

error: aborting due to previous error(s)
```

Is this perhaps a bug?

bors commented:
:umbrella: The latest upstream changes (presumably #42332) made this pull request unmergeable. Please resolve the merge conflicts.

arielb1 commented:
travis failure:
```
[01:04:16] error[E0061]: this function takes 3 parameters but 2 parameters were supplied

[01:04:16]    --> /checkout/src/libsyntax/util/parser_testing.rs:23:23

[01:04:16]     |

[01:04:16] 23  |       filemap_to_stream(&ps, ps.codemap().new_filemap("bogofile".to_string(), source_str))

[01:04:16]     |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 3 parameters

[01:04:16]     | 

[01:04:16]    ::: /checkout/src/libsyntax/parse/mod.rs

[01:04:16]     |

[01:04:16] 216 | / pub fn filemap_to_stream(sess: &ParseSess, filemap: Rc<FileMap>, override_span: Option<Span>)

[01:04:16] 217 | |                          -> TokenStream {

[01:04:16] 218 | |     let mut srdr = lexer::StringReader::new(sess, filemap);

[01:04:16] 219 | |     srdr.override_span = override_span;

[01:04:16] 220 | |     srdr.real_token();

[01:04:16] 221 | |     panictry!(srdr.parse_all_token_trees())

[01:04:16] 222 | | }

[01:04:16]     | |_- defined here

[01:04:16] 

[01:04:23] error[E0061]: this function takes 1 parameter but 0 parameters were supplied

[01:04:23]    --> /checkout/src/libsyntax/parse/mod.rs:690:40

[01:04:23]     |

[01:04:23] 690 |                           segments: vec![ast::PathSegment::crate_root(),

[01:04:23]     |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 1 parameter

[01:04:23]     | 

[01:04:23]    ::: /checkout/src/libsyntax/ast.rs

[01:04:23]     |

[01:04:23] 137 | /     pub fn crate_root(span: Span) -> Self {

[01:04:23] 138 | |         PathSegment {

[01:04:23] 139 | |             identifier: Ident { ctxt: span.ctxt, ..keywords::CrateRoot.ident() },

[01:04:23] 140 | |             span: span,

[01:04:23] 141 | |             parameters: None,

[01:04:23] 142 | |         }

[01:04:23] 143 | |     }

[01:04:23]     | |_____- defined here

[01:04:23] 

[01:04:28]    Compiling syntax_ext v0.0.0 (file:///checkout/src/libsyntax_ext)

[01:04:33] error: aborting due to previous error(s)
```

carols10cents commented:
Friendly ping @jseyfried to keep this on your radar! :)

bors commented:
:umbrella: The latest upstream changes (presumably #42712) made this pull request unmergeable. Please resolve the merge conflicts.

kennytm commented on src/tools/tidy/src/features.rs:
```diff
@@ -245,7 +245,7 @@ fn get_and_check_lib_features(base_src_path: &Path,
                         tidy_error!(bad, "{}:{}: {}", file.display(), line, msg);
                     };
-                    if lang_features.contains_key(name) {
+                    if lang_features.contains_key(name) && feature_name != "proc_macro" {
```
tidy literally failed.

```
[00:01:46]    Compiling tidy v0.1.0 (file:///checkout/src/tools/tidy)
[00:01:46] error[E0425]: cannot find value `feature_name` in this scope
[00:01:46]    --> /checkout/src/tools/tidy/src/features.rs:248:60
[00:01:46]     |
[00:01:46] 248 |                     if lang_features.contains_key(name) && feature_name != "proc_macro" {
[00:01:46]     |                                                            ^^^^^^^^^^^^ not found in this scope
[00:01:46] 
[00:01:47] error: aborting due to previous error
[00:01:47] 
[00:01:47] error: Could not compile `tidy`.
```

jseyfried replied:
rebase error :(

bors commented:
:umbrella: The latest upstream changes (presumably #42495) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
I've locally merged this to master and pushed up a few fixes as well at https://github.com/alexcrichton/rust/commits/proc_macro_api:

* https://github.com/alexcrichton/rust/commit/83a3ab2989d05dc3826587012e9db106503576e3 fixes the mention @kennytm pointed out above
* https://github.com/alexcrichton/rust/commit/bd3ee6b30bb0a7391d3647e3be8592275326492f seemed like it was just a minor fixup
* https://github.com/alexcrichton/rust/commit/9ad6d81aa8156a5e225a53a6e75bf3a0b48b5299 was the most interesting. I found that after rebasing this branch was panicking, but reverting the changes in these files seemed to resolve that (the panic was related to these changes). I don't actually know what these changes are, though, so it'd be good to get another set of eyes!
* https://github.com/alexcrichton/rust/commit/59476d4c3c69f62b50dc5c4660eb0f515c1d94f9 was just fixing some tests for some renamings

I'd be fine pushing these changes to this branch and r=nrc'ing, but https://github.com/alexcrichton/rust/commit/9ad6d81aa8156a5e225a53a6e75bf3a0b48b5299  was particularly worrisome.

@nrc mind taking a look at those changes? Would you be ok landing this if I pushed those changes to this PR?

jseyfried commented:
@alexcrichton 
Thanks! I believe your third bullet point is the only non-trivial thing left blocking this PR from merging -- I was working on figuring out the panics after rebase yesterday.

I believe the changes you reverted were necessary before the rebase, but I don't remember the details about why it was needed. If tests pass without the changes, we're probably fine for now. We need to refactor proc macro def ids at some point, and that will supersede the commit in question anyway.

I'll merge in your commits tonight and we should be ready to merge!

@nrc 
I've added and changed a fair amount of API as per @dtolnay's and other's suggestions (summarized in https://github.com/rust-lang/rust/pull/40939#issuecomment-303302410) -- this could use another review pass.

clarcharr commented:
I still think that it's very important to be able to parse literals as values, and that regardless of internal representation, it seems reasonable to not have to replicate that passing in downstream crates.

eddyb commented:
@clarcharr I'd much rather have people use a third-party crate to parse Rust syntactical elements such as expressions (or a subset thereof) and design the token system to be more orthogonal.

SimonSapin commented:
I see that `TokenTreeIter::Item` is `TokenTree`, not `&'a TokenTree`. Is there reference-counting involved? Do we want to commit to that?

I also second @clarcharr’s comment that `Literal` should have some way to extract the value without parsing `.to_string()`. Maybe methods like `Literal::as_char(&self) -> Option<char>` etc for each possible type? To keep it more extensible that an enum which would need to be exhaustive.

goffrie commented:
@eddyb But we're not talking about full expressions here, just single-token literals (`123_456u32`, `r##"ok#""`), which IMO should be part of the base API. Having to stringify these then reach out to an external crate seems pointless.

eddyb commented:
@goffrie Okay, yeah, I don't really think stringification should be required.

clarcharr commented:
re: what I said, this isn't about large syntactical things like expressions and items (the new macros were explicitly designed to push such parsing on to the crate ecosystem) but rather that if someone wants to extract the string from a literal, they have to re-parse the string to remove the escapes

It makes sense to be able to parse literals at least. Like I don't think that people should have to grab an extra crate to make a macro like thing!("one", "two", "three") work, and I especially don't want people who use ASCII only do benefit from a lack of parsing while forcing users of other languages to have to do so.

mystor commented:
I would generally agree that proc-macro should provide mechanisms for extracting values out of `Literal` objects, but I don't think that adding those methods should block landing this patch on nightly.

Until we get tools like that, you can use crates like my literalext crate to extract the values (https://github.com/mystor/literalext).

jseyfried commented:
@SimonSapin 
> I see that TokenTreeIter::Item is TokenTree, not &'a TokenTree. Is there reference-counting involved? Do we want to commit to that?

Yeah, `TokenTree` leans on ref-counting for efficiency. I believe we want to commit to that, otherwise it's a pretty major change to the structure and API of `TokenTree`. 

Right now `TokenTreeIter::Item` can't be a `&'a TokenTree` due to the architecture of `TokenTree`. It will be even less possible with @eddyb's proposed efficiency improvements through flattening trees in memory.

alexcrichton commented:
@jseyfried 

I ended up tracking down my [bug from before](https://github.com/rust-lang/rust/pull/40939#issuecomment-305898348) and it turns out to crop up [in the translation of `Interpolated`](https://github.com/jseyfried/rust/blob/a703056669d694b263f8d08341eaf0b1fda94894/src/libproc_macro/lib.rs#L504-L507). Notably we have no mechanism for translating a libsyntax AST to a `TokenStream` in a way that preserves all original spans (I believe), so this likely won't be an easy thing to fix.

Anyway, just wanted to let you know that it's probably not a quick fix here, and I'd be totally willing to help solve this bug after the PR lands!

nrc commented:
> I would generally agree that proc-macro should provide mechanisms for extracting values out of Literal objects, but I don't think that adding those methods should block landing this patch on nightly.

I agree - we're not going to get all the API right straight away and it will be unstable for a while. Better to land sooner and experiment than try and be complete.

nrc commented:
I had a look at the new commit from @jseyfried and he rebasing stuff from @alexcrichton. It mostly looks good to me. I'm not sure about https://github.com/alexcrichton/rust/commit/9ad6d81aa8156a5e225a53a6e75bf3a0b48b5299 but @jseyfried has ok'ed it in https://github.com/rust-lang/rust/pull/40939#issuecomment-310258473, so lets merge this thing.

jseyfried commented:
@bors r=nrc

bors commented:
:pushpin: Commit d316874 has been approved by `nrc`

<!-- @bors r=nrc d316874c87e25669895c306658e15aa3746d66ab -->

bors commented:
:hourglass: Testing commit d316874c87e25669895c306658e15aa3746d66ab with merge 392dce316b2375bc1db4f5ad7aa14811123642c0...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3709)

Mark-Simulacrum commented:
@bors retry -- presumably spurious failure

```
[00:59:07] failures:
[00:59:07]
[00:59:07] ---- [debuginfo-gdb] debuginfo-gdb\static-method-on-struct-and-enum.rs stdout ----
[00:59:07]  NOTE: compiletest thinks it is using GDB without native rust support
[00:59:07] NOTE: compiletest thinks it is using GDB version 7010001
[00:59:07]
[00:59:07] error: gdb failed to execute
[00:59:07] thread '[debuginfo-gdb] debuginfo-gdb\static-method-on-struct-and-enum.rs' panicked at 'explicit panic', src\tools\compiletest\src\runtest.rs:1650
[00:59:07] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:59:07]
[00:59:07]
[00:59:07] failures:
[00:59:07]     [debuginfo-gdb] debuginfo-gdb\static-method-on-struct-and-enum.rs
[00:59:07]
[00:59:07] test result: FAILED. 96 passed; 1 failed; 11 ignored; 0 measured; 0 filtered out
```

kennytm commented:
cc #42693 (Spurious `gdb failed to execute` on Windows).

alexcrichton commented:
@bors: retry

* https://github.com/rust-lang/rust/issues/42693

bors commented:
:hourglass: Testing commit d316874c87e25669895c306658e15aa3746d66ab with merge 90e21825495a9c8375608319ee9dbf5d8d29bcc9...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/247346801?utm_source=github_status&utm_medium=notification)

kennytm commented:
Should be legit. Many 3-hour timeout + 4 MB logs with `dist-` images when compiling `rls-span`. Lots of error logs of the form

```
[01:44:52] ERROR:rustc_save_analysis::dump_visitor: Mis-calculated spans for path '_serde::export::Result<Self, __D::Error>'. Found 1 spans, expected 3. Found spans:
[01:44:52] ERROR:rustc_save_analysis::dump_visitor:     'Deserialize' in /cargo/registry/src/github.com-1ecc6299db9ec823/rls-span-0.4.0/src/compiler.rs, line 16
[01:44:52] ERROR:rustc_save_analysis::dump_visitor:     master span: /cargo/registry/src/github.com-1ecc6299db9ec823/rls-span-0.4.0/src/compiler.rs:16:48: 16:59: `Deserialize`
[01:22:01] ERROR:rustc_save_analysis::dump_visitor: Mis-calculated spans for path '__D::Error'. Found 1 spans, expected 2. Found spans:
[01:22:01] ERROR:rustc_save_analysis::dump_visitor:     'Deserialize' in /cargo/registry/src/github.com-1ecc6299db9ec823/rls-span-0.4.0/src/compiler.rs, line 16
[01:22:01] ERROR:rustc_save_analysis::dump_visitor:     master span: /cargo/registry/src/github.com-1ecc6299db9ec823/rls-span-0.4.0/src/compiler.rs:16:48: 16:59: `Deserialize`
[01:22:01] ERROR:rustc_save_analysis::dump_visitor: Mis-calculated spans for path '_serde::export::Formatter'. Found 1 spans, expected 3. Found spans:
[01:22:01] ERROR:rustc_save_analysis::dump_visitor:     'Deserialize' in /cargo/registry/src/github.com-1ecc6299db9ec823/rls-span-0.4.0/src/compiler.rs, line 16
[01:22:01] ERROR:rustc_save_analysis::dump_visitor:     master span: /cargo/registry/src/github.com-1ecc6299db9ec823/rls-span-0.4.0/src/compiler.rs:16:48: 16:59: `Deserialize`
[01:22:01] ERROR:rustc_save_analysis::dump_visitor: Mis-calculated spans for path '_serde::export::fmt::Result'. Found 1 spans, expected 4. Found spans:
[01:22:01] ERROR:rustc_save_analysis::dump_visitor:     'Deserialize' in /cargo/registry/src/github.com-1ecc6299db9ec823/rls-span-0.4.0/src/compiler.rs, line 16
[01:22:01] ERROR:rustc_save_analysis::dump_visitor:     master span: /cargo/registry/src/github.com-1ecc6299db9ec823/rls-span-0.4.0/src/compiler.rs:16:48: 16:59: `Deserialize`
```

bors commented:
:hourglass: Testing commit d316874c87e25669895c306658e15aa3746d66ab with merge df0b593f53619982e3416616427756ca390d8992...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/247528177?utm_source=github_status&utm_medium=notification)

Mark-Simulacrum commented:
@bors retry

Timeouts across the board.

Mark-Simulacrum commented:
@bors r-

Still seeing https://github.com/rust-lang/rust/pull/40939#issuecomment-311256737; not sure actually why bors tested this again...

bors commented:
:umbrella: The latest upstream changes (presumably #42902) made this pull request unmergeable. Please resolve the merge conflicts.

aidanhs commented:
@jseyfried just a quick reminder that there are some test failures that need a look!

alexcrichton commented:
@bors r=nrc

bors commented:
:pushpin: Commit 78fdbfc has been approved by `nrc`

<!-- @bors r=nrc 78fdbfc4008b52bcce201fd589ed84d2abb0419d -->

alexcrichton commented:
I rebased this on master, fixed a conflict or two, and added https://github.com/rust-lang/rust/pull/40939/commits/78fdbfc4008b52bcce201fd589ed84d2abb0419d. According to @nrc we were only supposed to do `-Zsave-analysis` for libstd anyway, and it has the nice side effect of papering over the bugs seen on CI anyway.

bors commented:
:hourglass: Testing commit 78fdbfc4008b52bcce201fd589ed84d2abb0419d with merge 4d526e0d14b43a87627cd6aca6c6f71ad1e07b6e...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3816), [status-travis](https://travis-ci.org/rust-lang/rust/builds/250527490?utm_source=github_status&utm_medium=notification)
Approved by: nrc
Pushing 4d526e0d14b43a87627cd6aca6c6f71ad1e07b6e to master...

jseyfried commented:
@alexcrichton Awesome, thanks!

For the record, the failing test was because tokens from a macros 1.1-style proc-macro (i.e. via `FromStr for TokenStream`) now are given the proc-macro's call site span, as opposed to the proc-macros's call site span with extra expansion information before this PR.

Since we no longer have the extra expansion information, @nrc's `save-analysis` code tries to re-parses the call site span and expects it to match the proc-macro generated tokens.

I had to get rid of the expansion information for backwards compatibility -- if the expansion information showed that the span came from a procedural macro 2.0, then the underlying tokens would resolve hygienically (as per hygiene 2.0). This would be a backwards-incompatible change, and would mean that stringifying, reparsing and returning the input of a proc-macro would change its semantics, even if the input were entirely unexpanded.

If we want to fix this, we could instead mark tokens from `FromStr for TokenStream` specially so that save-analysis would know that they are "weird", but so that they would still behave like unexpanded tokens w.r.t. hygiene. I don't have a strong opinion here.

oli-obk commented:
The lack of expansion info broke clippy together with proc macros, since code inside derives will now trigger lints. Is there any new way to detect expanded code? If not, what needs to be done in order to create expansion info that doesn't influence save analysis

nrc commented:
> Since we no longer have the extra expansion information, @nrc's save-analysis code tries to re-parses the call site span and expects it to match the proc-macro generated tokens.

Ah, I wondered why I was seeing so many errors in the RLS :-(

I think we must fix this. This feels like another reason not to conflate macro hygiene with expansion history. That might be too hard a way to fix this though.

> If we want to fix this, we could instead mark tokens from FromStr for TokenStream specially so that save-analysis would know that they are "weird", but so that they would still behave like unexpanded tokens w.r.t. hygiene. I don't have a strong opinion here.

Not using spans for hygiene seems a much nicer fix here - the span should reflect the expansion history and the hygiene should be altered (to erase hygiene, essentially). I expect we want the correct expansion info for errors as well as save-analysis.

oli-obk commented:
I already have a fix in https://github.com/rust-lang/rust/pull/43179

jseyfried commented:
@nrc 
We could instead just include hygiene bending information along with expansion information so that span expansion info and hygiene info remain unified (@oli-obk's solution is fine for `impl FromStr for TokenStream`, so we're OK for now).

That is, I don't think think hygiene bending requires us to split up span/hygiene info or construct an "fake" expansion history just for hygiene. I think it'd be nicer to have a single source of truth for expansion info that is expressive enough for hygiene bending.

nrc commented:
So #43179 did not fix the expansion info issue (at least not completely) - I'm still seeing errors in the RLS where we are treating macro-generated names as hand-written ones (and I think this is causing #43371 too).
